### PR TITLE
feat: implement xlinectl `txn`, `watch` and `lock` commands

### PR DESCRIPTION
**Pull request by bsbds**
Based on: #427 
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline-client/src/types/auth.rs [hunk 1]
```diff
@@ -213,7 +213,7 @@ impl From<AuthUserRevokeRoleRequest> for xlineapi::AuthUserRevokeRoleRequest {
 }
 
 /// Request for `AuthRoleAdd`
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (Phoenix500526) [inline xline-client/src/types/auth.rs:216]: What's this `PartialEq` for?

**Review by Phoenix500526 — COMMENTED** [review]
LGTM

**Review by Phoenix500526 — DISMISSED** [review]
LGTM

**Comment by bsbds** [comment]
rebased

File: xlinectl/src/command/watch.rs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            };
+            let key = matches.get_one::<String>("key").expect("required");
+            let range_end = matches.get_one::<String>("range_end");
+            let req = build_request(&matches)(key, range_end.map(String::as_str));
+            assert_eq!(Some(req), self.req);
+        }
+    }
+
+    #[test]
+    fn valid() {
```
Comment (Phoenix500526) [inline xlinectl/src/command/watch.rs:190]: Same as above.

File: xlinectl/src/command/txn.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        let input = io::read_to_string(io::stdin())?;
+        let args: Vec<_> = input
+            .split("\n\n")
+            .filter(|s| !s.trim().is_empty())
+            .map(ToOwned::to_owned)
+            .collect();
+        if args.len() != 3 {
+            bail!("the arg length should be 3");
+        }
+        #[allow(clippy::indexing_slicing)] // checked above so its safe to index
```
Comment (Phoenix500526) [inline xlinectl/src/command/txn.rs:54]: typo. "it's" not "its"

File: xlinectl/src/command/txn.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+/// Parse one line of compare command
+fn parse_cmp_line(line: &str) -> Result<Compare> {
+    // match something like `mod("key1) > "0"`
+    #[allow(clippy::unwrap_used)] // This regex is tested to be valid
+    let re = Regex::new(r#"(\w+)\("([^"]+)"\) ([<=>]) "([^"]+)"$"#).unwrap();
+
+    match re.captures(line) {
+        #[allow(clippy::indexing_slicing)] // checked in regex so its safe to index
```
Comment (Phoenix500526) [inline xlinectl/src/command/txn.rs:88]: Same as above

File: xlinectl/src/command/lock.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+pub(crate) fn command() -> Command {
+    Command::new("lock")
+        .about("Acquire a lock, which will return a unique key that exists so long as the lock is held")
+        .arg(arg!(<lockname> "name of the lock"))
+}
+
+/// Build request from matches
+pub(crate) fn build_request(matches: &ArgMatches) -> LockRequest {
+    let name = matches.get_one::<String>("lockname").expect("required");
+    LockRequest::new().with_name(name.as_bytes())
```
Comment (themanforfree) [inline xlinectl/src/command/lock.rs:21]: Shell we change this to `LockRequest::new(name.as_bytes())`

File: xlinectl/src/command/lock.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    let resp = client.lock_client().lock(req).await?;
+
+    resp.print();
+
+    signal::ctrl_c().await.expect("failed to listen for event");
+
+    println!("releasing the lock");
+
+    let unlock_req = UnlockRequest::new().with_key(resp.key);
```
Comment (themanforfree) [inline xlinectl/src/command/lock.rs:36]: Same as above

File: xlinectl/src/command/watch.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+};
+
+use crate::utils::printer::Printer;
+
+/// Definition of `watch` command
+pub(crate) fn command() -> Command {
+    Command::new("watch")
+        .about("Watches events stream on keys or prefixes")
+        .arg(
+            arg!([key] "The key or preifx, not needed in interactive mode")
```
Comment (themanforfree) [inline xlinectl/src/command/watch.rs:18]: typo, `prefix`

File: xlinectl/src/command/lock.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_case_struct;
+
+    test_case_struct!(LockRequest);
+
+    #[test]
+    fn valid() {
```
[the remark this answers is not in the document]
Comment (bsbds) [reply xlinectl/src/command/lock.rs]: done

File: xlinectl/src/command/txn.rs (hunk 3 above)
Comment (bsbds) [reply xlinectl/src/command/txn.rs]: fixed

File: xlinectl/src/command/lock.rs (hunk 5 above)
Comment (bsbds) [reply xlinectl/src/command/lock.rs]: done

File: xlinectl/src/command/watch.rs (hunk 7 above)
Comment (bsbds) [reply xlinectl/src/command/watch.rs]: fixed

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by themanforfree — APPROVED** [review]
LGTM

**Pull request merged**